### PR TITLE
Add `tctl` notifications commands

### PR DESCRIPTION
**Pull request by rudream**
## Purpose

Part of #37704 

This PR adds various `tctl` commands for creating and managing user-created notifications.

Admins can create notifications targeting one or more roles, or target a specific user. They can also list user-created user-specific notifications for a user, or user-created global notifications.

changelog: tctl allows cluster administrators to create custom notifications targeting Teleport users.

### CODE REVIEW

File: api/proto/teleport/notifications/v1/notifications_service.proto [hunk 1]
```diff
@@ -109,3 +114,21 @@ message UpsertUserLastSeenNotificationRequest {
...
+message ListAllUserCreatedNotificationsForUserRequest {
+  // page_size is the size of the page to return.
+  int32 page_size = 1;
+  // page_token is the next_page_token value returned from a previous ListUserNotifications request, if any.
+  string page_token = 2;
+  // username is the username of the user the notifications being listed are for.
+  string username = 3;
+}
+
+// ListAllUserCreatedGlobalNotificationsRequest is the request for listing all user-created global notifications..
```
Comment (zmb3) [inline api/proto/teleport/notifications/v1/notifications_service.proto:128]: What is a user-created global notification?

File: lib/auth/notifications/notificationsv1/service.go [hunk 2]
```diff
@@ -444,3 +445,174 @@ func (s *Service) UpsertUserLastSeenNotification(ctx context.Context, req *notif
...
+	return out, nil
+}
+
+// DeleteGlobalNotification deletes a global notification.
+func (s *Service) DeleteGlobalNotification(ctx context.Context, req *notificationsv1.DeleteGlobalNotificationRequest) (*emptypb.Empty, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```
Comment (zmb3) [inline lib/auth/notifications/notificationsv1/service.go:501]: Creating a notification is an admin action but deleting one is not?

File: lib/auth/notifications/notificationsv1/service.go [hunk 3]
```diff
@@ -444,3 +445,174 @@ func (s *Service) UpsertUserLastSeenNotification(ctx context.Context, req *notif
...
+// ListAllUserCreatedNotificationsForUser returns a paginated list of all user-created user-specific notifications for a user. This should only be used by admins.
+func (s *Service) ListAllUserCreatedNotificationsForUser(ctx context.Context, req *notificationsv1.ListAllUserCreatedNotificationsForUserRequest) (*notificationsv1.ListNotificationsResponse, error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !authz.HasBuiltinRole(*authCtx, string(types.RoleAdmin)) {
+		return nil, trace.AccessDenied("only users with the admin role can list notifications for another user")
```
Comment (zmb3) [inline lib/auth/notifications/notificationsv1/service.go:533]: Technically `RoleAdmin` cannot be assigned to users, so this message is a bit confusing.

File: lib/auth/notifications/notificationsv1/service.go [hunk 4]
```diff
@@ -444,3 +445,174 @@ func (s *Service) UpsertUserLastSeenNotification(ctx context.Context, req *notif
...
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindNotification, types.VerbList); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if req.Username == "" {
```
Comment (zmb3) [inline lib/auth/notifications/notificationsv1/service.go:540]: Should you do this check first before you start doing RBAC checks?

File: lib/web/ui/labels.go [hunk 5]
```diff
@@ -51,6 +51,30 @@ func makeLabels(labelMaps ...map[string]string) []Label {
 	return labels
 }
 
+// makeLabelsIncludingInternalis a function that transforms map[string]string arguments passed to it to sorted slice of Labels, including Teleport internal labels.
```
Comment (zmb3) [inline lib/web/ui/labels.go:54]:
```suggestion
// makeLabelsIncludingInternal transforms map[string]string arguments passed to it to sorted slice of Labels, including Teleport internal labels.
```

File: tool/tctl/common/notification_command.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+					Username: n.user,
+				},
+			},
+		})
+
+		if err != nil {
+			return trail.FromGRPC(err)
+		}
+
+		fmt.Printf("\n\ncreated notification: \n%v\n", created)
```
Comment (zmb3) [inline tool/tctl/common/notification_command.go:150]: Remove this?

File: tool/tctl/common/notification_command.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+					Notification: &notificationspb.Notification{
+						Kind:     types.KindNotification,
+						SubKind:  subKind,
+						Metadata: meta,
+						Spec:     &notificationspb.NotificationSpec{},
+					},
+				},
+			},
+		})
+
```
Comment (zmb3) [inline tool/tctl/common/notification_command.go:176]:
```suggestion
```

Keep the error check in the same logical block as the code that generated the error.

File: tool/tctl/common/notification_command.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+	}
+
+	displayNotifications(n.format, result)
+	return nil
+}
+
+func displayNotifications(format string, notifications []*notificationspb.Notification) {
+	switch format {
+	case teleport.Text:
+		var rows [][]string
```
Comment (zmb3) [inline tool/tctl/common/notification_command.go:261]:
```suggestion
		rows := make([][]string, 0, len(notifications))
```

File: tool/tctl/common/notification_command.go [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+	default:
+		_, err = client.DeleteGlobalNotification(ctx, &notificationspb.DeleteGlobalNotificationRequest{
+			NotificationId: n.title,
+		})
+	}
+
+	return trail.FromGRPC(err)
+}
+
+func (n *NotificationCommand) splitRoles() []string {
```
Comment (zmb3) [inline tool/tctl/common/notification_command.go:299]: does kingpin not handle this for us?

File: web/packages/teleport/src/Notifications/Notification.tsx [hunk 10]
```diff
@@ -96,6 +96,27 @@ export function Notification({
     hideNotificationAttempt.status === 'processing' ||
     (view === 'Unread' && clicked)
   ) {
+    // If this is a text content notification, the dialog should still be renderable. This is to prevent the text content dialog immediately disappearing
```
Comment (zmb3) [inline web/packages/teleport/src/Notifications/Notification.tsx:99]: This doesn't seem related to tctl..

File: api/proto/teleport/notifications/v1/notifications_service.proto (hunk 1 above)
Comment (rudream) [reply api/proto/teleport/notifications/v1/notifications_service.proto:128]: Any global notification which was created by a user using `tctl notifications create` (as opposed to global notifications generated by Teleport, eg. access requests). The main use of the `tctl notifications` interface is to create and manage custom notifications, so I didn't think it would be necessary to include _all_ global notifications as that would clutter the list with dozens of irrelevant notifications.

File: lib/auth/notifications/notificationsv1/service.go (hunk 2 above)
Comment (rudream) [reply lib/auth/notifications/notificationsv1/service.go:501]: I did this because the reasoning given in the RFD for making creating a notification an admin action was just to mitigate the possibility of a bad actor creating a malicious notification to phish users, but I'll make deleting one an admin action as well

File: web/packages/teleport/src/Notifications/Notification.tsx (hunk 10 above)
Comment (rudream) [reply web/packages/teleport/src/Notifications/Notification.tsx:99]: This is relevant because this fixes a bug causing text content notifications to be unreadable when opened from the "unread" tab, and text content notifications can only ever be generated by `tctl notifications create`. The UI part of this feature would be broken without this

File: api/proto/teleport/notifications/v1/notifications_service.proto (hunk 1 above)
Comment (zmb3) [reply api/proto/teleport/notifications/v1/notifications_service.proto:128]: If there are dozens of global notifications that would clutter this list, wouldn't that also be cluttering every user's notification list in the UI?

`tctl` is an admin tool more than it is a user tool. I think its acceptable (and preferred) for an admin tool to show you everything rather than hide stuff from you.

Comment (rudream) [reply api/proto/teleport/notifications/v1/notifications_service.proto:128]: > wouldn't that also be cluttering every user's notification list in the UI?

No, since for regular users in the UI, only global notifications which match the user would be shown to them, ie. for access request notifications, a user in the UI will only see the notifications relevant to them (that their request was approved/denied or that a request is pending their review)

File: lib/web/ui/notification.go [hunk 11]
```diff
@@ -36,7 +36,7 @@ type Notification struct {
 
 // MakeNotification creates a notification object for the WebUI.
 func MakeNotification(notification *notificationsv1.Notification) Notification {
-	labels := makeLabels(notification.Metadata.Labels)
+	labels := makeLabelsIncludingInternal(notification.Metadata.Labels)
```
[the remark this answers is not in the document]
Comment (rudream) [reply lib/web/ui/notification.go:39]: This is because for notifications, the notification text content and a few other user-facing values are stored in these labels, and they're required when building the response to send to the UI so it knows what to display

File: tool/tctl/common/notification_command.go [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+	var pageToken string
+	for {
+		var resp *notificationspb.ListNotificationsResponse
+		var err error
+
+		// If a user was specified, list user-created user-specific notifications for them, if not, default to listing user-created global notifications.
+		if n.user != "" {
+			resp, err = client.ListAllUserCreatedNotificationsForUser(ctx, &notificationspb.ListAllUserCreatedNotificationsForUserRequest{
+				Username:  n.user,
+				PageSize:  512,
```
[the remark this answers is not in the document]
Comment (rudream) [reply tool/tctl/common/notification_command.go:227]: Looks like this was a temporary leftover

File: api/proto/teleport/notifications/v1/notifications_service.proto [hunk 13]
```diff
@@ -41,6 +41,11 @@ service NotificationService {
   // ListNotifications returns a paginated list of a user's notifications.
   rpc ListNotifications(ListNotificationsRequest) returns (ListNotificationsResponse);
 
+  // ListAllUserCreatedNotificationsForUser returns a paginated list of all user-created user-specific notifications for a user. This should only be used by admins.
+  rpc ListAllUserCreatedNotificationsForUser(ListAllUserCreatedNotificationsForUserRequest) returns (ListNotificationsResponse);
+  // ListAllUserCreatedGlobalNotifications returns a paginated list of all user-created global notifications. This should only be used by admins.
+  rpc ListAllUserCreatedGlobalNotifications(ListAllUserCreatedGlobalNotificationsRequest) returns (ListNotificationsResponse);
```
[the remark this answers is not in the document]
Comment (rudream) [reply api/proto/teleport/notifications/v1/notifications_service.proto:47]: I tried this initially, but due to all the possible options `ListNotifications` would need to handle (regular listing notifications, listing only user-created user-specific notifications, listing only user-created global notifications, listing all global notifications) it made the request complicated and hard to follow, and it also made the logic in `ListNotifications` very cluttered and hard to follow. Splitting it up into these dedicated RPCs makes the code much more modular and the methods easier for consumers to use.

File: tool/tctl/common/notification_command.go [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+	n.create = notif.Command("create", "Create a cluster notification.").Alias("add")
+	n.create.Flag("user", "Target a specific user.").StringVar(&n.user)
+	n.create.Flag("roles", "Target a specific set of roles. By default, this will target all users with any of the provided roles, use --require-all-roles to exclusively target users with all of them.").StringsVar(&n.roles)
+	n.create.Flag("require-all-roles", "Set whether this notification should target users who have all of the provided roles.").BoolVar(&n.requireAllRoles)
+	n.create.Flag("title", "Set the notification's title.").Short('t').Required().StringVar(&n.title)
+	n.create.Flag("content", "Set the notification's content.").Required().StringVar(&n.content)
+	n.create.Flag("warning", "Set whether this notification is a warning notification.").BoolVar(&n.warning)
+
+	n.ls = notif.Command("ls", "List user-created cluster notifications. This will list all user-created global notifications by default, to list user-specific notifications for a user, use the --user flag.")
+	n.ls.Flag("user", "Set which user to list user-created user-specific notifications for.").StringVar(&n.user)
```
[the remark this answers is not in the document]
Comment (rudream) [reply tool/tctl/common/notification_command.go:75]: Any notification that was created using `tctl` and targeted at a specific user. I'll make this more clear

File: lib/web/ui/notification.go [hunk 15]
```diff
@@ -26,12 +26,13 @@ import (
 )
 
 type Notification struct {
-	Id      string    `json:"id"`
-	Title   string    `json:"title"`
-	SubKind string    `json:"subKind"`
-	Created time.Time `json:"created"`
-	Clicked bool      `json:"clicked"`
-	Labels  []Label   `json:"labels"`
+	Id          string    `json:"id"`
```
Comment (zmb3) [inline lib/web/ui/notification.go:29]:
```suggestion
	ID          string    `json:"id"`
```

File: tool/tctl/common/notification_command.go [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+	case teleport.YAML:
+		utils.WriteYAML(w, notifications)
+	default:
+		// Do nothing, kingpin validates the --format flag before we ever get here.
+	}
+}
+
+// Remove removes a notification.
+func (n *NotificationCommand) Remove(ctx context.Context, client *authclient.Client) error {
+	var err error
```
Comment (zmb3) [inline tool/tctl/common/notification_command.go:290]:
```suggestion
```

**Review by zmb3 — APPROVED** [review]
Please make sure to squash and rebase so we get rid of all the "CR" in the commit message.

File: tool/tctl/common/notification_command.go [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+	}
+	return true, trace.Wrap(err)
+}
+
+// Create creates a new notification.
+func (n *NotificationCommand) Create(ctx context.Context, client *authclient.Client) error {
+	meta := &headerv1.Metadata{
+		Labels: map[string]string{
+			types.NotificationTitleLabel:       n.title,
+			types.NotificationTextContentLabel: n.content,
```
[the remark this answers is not in the document]
Comment (rudream) [reply tool/tctl/common/notification_command.go:113]: We did this so we could use the `Metadata` struct to contain all of a notification's metadata, since every unique type of notification will have completely different metadata fields. For example, this `TextContent` field exists only for user-created notifications, and no other notification type will have it. The only field that is universal among all notifications is `Title`, which could be in `Spec` but for the sake of consistency and keeping all metadata grouped together in one place we decided to store it here alongside all other notification metadata.

**Review by rosstimothy — APPROVED** [review]

File: tool/tctl/common/notification_command_test.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+	// Test creating a global notification for users with the test-1 role.
+	buf, err = runNotificationsCommand(t, clt, []string{"create", "--roles", "test-1", "--title", "test-1 notification", "--content", "This is a test notification."})
+	require.NoError(t, err)
+	require.Contains(t, buf.String(), "for users with one or more of the following roles: [test-1]")
+	globalNotificationId := strings.Split(buf.String(), " ")[2]
+
+	// We periodically check with a timeout since it can take some time for the item to be replicated in the cache and be available for listing.
+	require.EventuallyWithT(t, func(collectT *assert.CollectT) {
+		// List notifications for auditor and verify that auditor notification exists.
+		buf, err = runNotificationsCommand(t, clt, []string{"ls", "--user", auditorUsername})
```
Comment (zmb3) [inline tool/tctl/common/notification_command_test.go:75]: If this fails the `t`, then the whole test will fail instead of only failing the `EventuallyWithT`.

I think that's probably fine in this case (making `runNotificationsCommand` work with a `testing.T` or a `assert.CollectT` is probably a bigger change), but wanted to point it out.

**Review by zmb3 — APPROVED** [review]
